packages/aws-cdk-lib/aws-ssm/lib/parameter.ts

```diff
@@ -473,9 +473,15 @@ export class StringParameter extends ParameterBase implements IStringParameter {
 
     const type = attrs.type ?? attrs.valueType ?? ParameterValueType.STRING;
 
-    const stringValue = attrs.version
-      ? new CfnDynamicReference(CfnDynamicReferenceService.SSM, `${attrs.parameterName}:${Tokenization.stringifyNumber(attrs.version)}`).toString()
-      : new CfnParameter(scope, `${id}.Parameter`, { type: `AWS::SSM::Parameter::Value<${type}>`, default: attrs.parameterName }).valueAsString;
+    let stringValue: string;
+    if (attrs.version) {
+      stringValue = new CfnDynamicReference(CfnDynamicReferenceService.SSM, `${attrs.parameterName}:${Tokenization.stringifyNumber(attrs.version)}`).toString();
+    } else if (Token.isUnresolved(attrs.parameterName)) {
```

Comment:
I think there are cases where this could be a breaking change. There could be cases where `Token.isUnresolved` returns `true`, but the value in the template is still a resolved string.

For example

```ts
const parameterName = Lazy.string({ produce: () => 'myParameterName' });
```

`Token.isUnresolved(parameterName)` will return true, but the value that will be output will be 

```json
{
  "Type": "AWS::SSM::Parameter::Value<STRING>",
  "Default": "myParameterName",
}
```

We might need to use `Tokenization.isResolvable` instead.

Reply:
Hi @corymhall thanks! I confirmed you are right (I added a test case and it failed.)

Now I tried `Tokenization.isResolvable` but for some reason it was not evaluated as `true` for a cross stack reference. I'll dive deep the reason later.

Perhaps we need to resolve the token first, and check if it contains CFn intrinsics, as you've done on `cfn-resources.ts`.

Reply:
`Tokenization.isResolvable(attrs.parameterName)` won't work as expected because `typeof(attrs.parameterName)` is always `'string'`, which fails this evaluation.

https://github.com/aws/aws-cdk/blob/1d7a9a80b08d41ce8759bed9286adaa8259c2bc8/packages/aws-cdk-lib/core/lib/token.ts#L319-L321


So to determine whether a parameter name can be resolved to a concrete string, we can actually resolve it. Like this:

```ts
if (typeof Stack.of(scope).resolve(attrs.parameterName) != 'string') 
```

However, if we resolve the token here, it means a lazy string isn't actually evaluated lazily enough. Instead it's evaluated immediately to determine the condition. I'm not sure if it will break something though. A similar approach is also used in s3-deployment module, so maybe it's just fine.

https://github.com/aws/aws-cdk/blob/1d7a9a80b08d41ce8759bed9286adaa8259c2bc8/packages/aws-cdk-lib/aws-s3-deployment/lib/render-data.ts#L18-L19

Alternatively, we can create a method like `isLazyString` to determine if a string is not a lazy value without resolving it.  But then we cannot cover the case when `Lazy.string` produces cfn intrinsics like this:

```ts
Lazy.string({ produce: () => someResource.ref });
```

Another consideration is if we should really care this edge case. I think almost all of the devs don't use a Lazy string as a ssm parameter name to import it. At least we can fix it if someone would report an issue.

Hope you can suggest which way we should go, thanks! @corymhall

Reply:
It seems like what we are really trying to do is determine if the value is an intrinsic or not. In that case I think it might be best to just add a  `isIntrinsic` method to [Intrinsic](https://github.com/aws/aws-cdk/blob/53d24f1a1357410746052bbf25b5b89b985f7b80/packages/aws-cdk-lib/core/lib/private/intrinsic.ts#L36-L36)

Reply:
Thanks for the insight! I agree with that the only case we have to handle is when parameterName contains intrinsics.

But `attrs.parameterName` is just a string (like `${Token[TOKEN.274]}`) before resolution. Is it possible to determine if the string is actually an instrinsic without resolving it?

Reply:
It seems we can get backing tokens from a string by `TokenMap.instance().splitString(string).tokens` without actually resolving it. On second thought, however, if we also want to handle the case `Lazy.string({ produce: () => someResource.ref })`, we have to resolve the token anyway to determine if it's an intrinsic. But the disadvantage is that it has a side effect of resolving a lazy token in an unexpected timing.

@corymhall Do you think we really have to handle the lazy token edge case? I don't think it is a critical problem and is worth complicating the implementation, and would rather go with the original patch. To avoid a breaking change, maybe we can just add a feature flag.

Reply:
@tmokmss yeah there is no way to handle the `Lazy` case so lets let that fallback to the old behavior. Lets take an approach where we explicitly cover certain scenarios and everything else just falls back to the existing functionality. The one I can think of is an `isIntrinsic`

```ts
Token.isUnresolved(attrs.parameterName) && Intrinsic.isInstrinsic(Tokenization.reverseString(attrs.parameterName).firstToken)

```

Reply:
@corymhall Thanks. I tried it and found that it won't work as expected :(
Because in the test case below, the parameterName is also a `Intrinsic` (to be precise, it's a `Reference` that extends `Intrinsic`), failing the test.

https://github.com/aws/aws-cdk/blob/54bcad2b172946e7879e0296b19c656a4716eca7/packages/aws-cdk-lib/aws-ssm/test/parameter.test.ts#L1045-L1055

So I guess what we need is a way to
1. determine if the token contains a CFn intrinsic
2. without resolving the token, to avoid side effects to Lazy tokens

I'll try to find how to achieve this, but let me know if you have a good idea.

Reply:
What if we just don't handle the `Reference` case and only handle where `isIntrinsic() && !isReference()`?

Reply:
@corymhall Thanks. I confirmed it won't work either. The `parameterName` is actually also a `Reference` in other test cases like `When a parameter name contains tokens, use dynamic reference instead` too. I guess it's because of this function:

https://github.com/aws/aws-cdk/blob/c135656bb0b6de9cce639218a83acf958f9bca4e/packages/aws-cdk-lib/core/lib/resource.ts#L254

To get the underlying token, it seems we must first resolve the reference, because the intrinsic is lazily produced. 

I'm now wondering whether switching a [parameter](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#aws-ssm-parameter-types) to a [dynamic reference](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html) should really be considered as a breaking change. As far as I read the docs, there seems to be no remarkable difference between them. Given it's also very rare to use a lazy token for parameter names, we can tolerate the change, maybe under a feature flag.

Reply:
`When a parameter name contains tokens, use dynamic reference instead` that case is what I am saying that we should not cover. It should just fall back to the old behavior and will not work in some cases (like before). Something like this should work though

```ts
ssm.StringParameter.fromStringParameterAttributes(this, 'param', {
  parameterName: Fn.getAtt(), // or Fn.conditionOr(), etc.
})
```

Reply:
There are some limitations https://github.com/aws/aws-cdk/pull/22239#issuecomment-1262069499

Reply:
@corymhall Thanks, I didn't know that! 

So how about letting users choose which they use, parameter or dynamic reference? We'll add a property like `forceDynamicReference?: boolean` (default to false) to `CommonStringParameterAttributes`. This is kind of a leaky abstraction, but it should at least solve all the problem above.

Reply:
@tmokmss lets do both. Lets cover what we can (my comment above) and then allow the user to force the dynamic reference. Neither should need a feature flag.

Reply:
@corymhall Now dynamic reference is used only when `cdk.Fn` is explicitly used. It can address the original use case reported on the issue. 

Other parameter name containing a token, we cannot decide if it contains CFn intrinsics or not without resolving the token, which is not ideal for lazy tokens. In that case, we just use CfnParameter, and letting a user to force dynamic reference by passing a flag.

I made the change to only `fromStringParameterAttributes`, not `fromListParameterAttributes`. I think we can do this in another PR.